6.2c  : make the ComponenfFiller fill a master or classic problem

guilpier-code opened the pull request:
Fixes [ticket ANT-3913](https://gopro-tickets.rte-france.com/browse/ANT-3913)

What's done : 
- [x] code cleaning
- [x] add a filter on variables to inject into the problem
- [x] add a filter on objectives to inject into the problem
- [x] unit tests on component filler for previous added code

payetvin committed:
Add optim config header

payetvin committed:
Add cpp and cmakelists

payetvin committed:
Add unit tests

payetvin committed:
format

payetvin committed:
Add constructors

payetvin committed:
format

payetvin committed:
Add getter

payetvin committed:
fix unit tests

payetvin committed:
use antares exception, rm dependency to yuni

payetvin committed:
remove uint

payetvin committed:
Remove libraries from optim config

payetvin committed:
link fmt

payetvin committed:
code smell

payetvin committed:
Fix include path

JasonMarechal25 committed:
Merge branch 'develop' into feature/6.2-objects

payetvin committed:
move files to a dir investment

payetvin committed:
Revert "move files to a dir investment"

This reverts commit 6d4bd12db16948b8dbb2c1879f93cdc035579776.

payetvin committed:
add location to variable

payetvin committed:
conflicts

payetvin committed:
Format

payetvin committed:
link correct lib

payetvin committed:
comments

payetvin committed:
format

Commit:
6.2-c : small simplifications

Commit:
6.2-c : more small simplifications

Commit:
6.2-c : move class VariablesBulkAddition declaration only where it's needed

Commit:
6.2-c : remove useless things from the component filler header

Commit:
6.2-c : adding a filter in variables before loading them in the problem

guilpier-code commented on src/solver/optim-model-filler/include/antares/solver/optim-model-filler/ComponentFiller.h:
```diff
@@ -23,25 +23,15 @@
 
 #include "ReadLinearConstraintVisitor.h"
 
-namespace Antares::ModelerStudy::SystemModel
-{
-class Component;
-class Variable;
-} // namespace Antares::ModelerStudy::SystemModel
-
-namespace Antares::Expressions::Visitors
-{
-class EvalVisitor;
-}
```
These forward declarations are useless. They are removed.

guilpier-code commented on src/solver/optim-model-filler/include/antares/solver/optim-model-filler/ComponentFiller.h:
```diff
@@ -79,49 +68,6 @@ class ComponentFiller: public LinearProblemApi::LinearProblemFiller
     const ScenarioGroupRepository& scenarioGroupRepository_;
-};
-
-class VariablesBulkAddition
```
**class VariablesBulkAddition** is required only for `ComponentFiller::addVariables(...)`'s own private usage. 
So it's moved to **ComponentFiller.cpp**

Commit:
[skip ci] 6.2-c : add a method addVariablesToMaster(...) to class ComponentFiller

Commit:
6.2-c : a bit of clarifications in unit tests about component filler

Commit:
6.2-c : fix a compilation error

Commit:
6.2-c : clarifications

Commit:
Merge remote-tracking branch 'remotes/origin/develop' into feature/6.2c--step-1

Commit:
6.2-c : setting unit tests environment for filling a master problem

Commit:
6.2-c : adding a unit test on filling a master linear problem

guilpier-code commented on src/solver/optim-model-filler/ComponentFiller.cpp:
```diff
@@ -49,9 +47,55 @@ std::optional<T> buildOptional(bool condition, T value)
 namespace Antares::Optimisation
 {
+
+class VariablesBulkAddition
```
`class VariablesBulkAddition` was moved here from **ComponentFiller.h**

guilpier-code commented on src/tests/end-to-end/modeler/test-modeler.cpp:
Just a bit of cleaning, no real change here.

guilpier-code commented on src/tests/inmemory-modeler/include/inmemory-modeler.h:
Just a bit of cleaning, no real change here.

guilpier-code commented on src/tests/inmemory-modeler/inmemory-modeler.cpp:
Just a bit of cleaning, no real change here.

guilpier-code commented on src/tests/src/io/outputs/testSimulationTable.cpp:
Just a bit of cleaning, no real change here.

guilpier-code commented on src/tests/src/solver/optim-model-filler/test_componentFiller.cpp:
Just a bit of cleaning, no real change here.

Commit:
6.2-c : clarify the unit test title + code format

Commit:
6.2-c : introducing a fixture to unit tests (common to other tests to come)

Commit:
6.2-c : adding a new unit tests on variables to be added on master problem/subproblem

Commit:
6.2-c : simplify loop over constraints

Commit:
6.2-c : objectives - offering possibility to add master / subproblem objectives to a linear problem

Commit:
6.2-c : introduce a abstract variable creator for the need of testing

Commit:
6.2-c : a bit of cleaning

Commit:
6.2-c : mainly fix a bug

Commit:
6.2-c - unit tests : introducing a objectives creator, to be injected in the fixture

Commit:
6.2-c - unit tests : adding a unit test on adding objective to problem depending on which problem (master/subproblem)

Commit:
6.2-c - unit tests : adding another test on adding objective to problem depending on which problem (master/subproblem)

flomnes commented on src/tests/src/solver/optim-model-filler/test-component-filler-to-master-pb.cpp:
```diff
@@ -0,0 +1,339 @@
+    }
+
+    void initialize(std::unique_ptr<VariablesCreator> varCreator,
+                    std::unique_ptr<ObjectivesCreator> objCreator)
```
Add template parameters, remove `initialize` method

```suggestion
template<class VariablesCreator, class ObjectivesCreator>
struct FactoryFixture
{
    FactoryFixture():
        linear_pb(false, "sirius"),
        optimEntityContainer(linear_pb, &dummy_data, &scenario_group_repo)
    {
    }

    void initialize(std::unique_ptr<VariablesCreator> varCreator,
                    std::unique_ptr<ObjectivesCreator> objCreator)
```

flomnes commented on src/tests/src/solver/optim-model-filler/test-component-filler-to-master-pb.cpp:
```diff
@@ -0,0 +1,339 @@
+
+BOOST_AUTO_TEST_SUITE_END()
+
+BOOST_FIXTURE_TEST_SUITE(add_constraints_to_master_linear_problem, FactoryFixture)
```
Move to 
```cpp
BOOST_FIXTURE_TEST_CASE(my_test_case, StudyFixture)
```
otherwise the fixture object is re-used in all tests in a test suite.

guilpier-code replied:
Done

a-zakir commented on src/solver/optim-model-filler/ComponentFiller.cpp:
```diff
@@ -336,13 +390,20 @@ void ComponentFiller::addConstraints(const LinearProblemApi::FillContext& ctx)
     }
 }
 
+void ComponentFiller::addObjectivesToMaster(const LinearProblemApi::FillContext& ctx)
```
This approach works, assuming callers will use either addObjectives or addObjectivesToMaster. However, it adds complexity by enriching ComponentFiller with filters, and even then, you still need a boolean flag to distinguish between the two cases.
Here’s how I would have approached it:

```
class ComponentFiller{
ComponentFiller::ComponentFiller(const ModelerStudy::SystemModel::Component& component,
                                 OptimEntityContainer& optimEntityContainer,
                                 const ScenarioGroupRepository& scenarioGroupRepository):
    component_(component),
    optimEntityContainer_(optimEntityContainer),
    scenarioGroupRepository_(scenarioGroupRepository),
    objectivesFilter_(subproblemObjectives)
{
switch (optimEntityContainer.**ProblemType()**){
  case Subproblems:
  variablesFilter_  = subproblemVariables ;
  ...
}
}
};
```
without addVariablesToMaster and addObjectivesToMaster

guilpier-code replied:
ok, please see [this comment ](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/3166#issuecomment-3437181648)from @flomnes

flomnes replied:
This was fixed by adding a Location argument to ComponentFiller's constructor

Commit:
6.2-c - unit tests : move test material to new source files

Commit:
6.2-c - unit tests : remove useless things

Commit:
6.2-c - unit tests : split utils into couples *.h/*.cpp

Commit:
6.2-c - unit tests : code formatting

flomnes commented:
As discussed with @tbittar , we shouldn't have specific methods for the master problem. Instead, add a `location` attribute to the `Antares::Optimisation::LinearProblemApi::FillContext` class and compare it with the variable/objective location

You may define an `constexpr bool operator<=(Location lhs, Location rhs)` on the location with the following rules
- x <= x for all x
- master_and_subproblems <= subproblems 
- master_and_subproblems <= master
- the rest is false

Commit:
6.2-c - unit tests : changes subsequent to code review

Commit:
Merge remote-tracking branch 'remotes/origin/develop' into feature/6.2c--step-1

Commit:
6.2-c : rename modeler.solve() into modeler.run()

Commit:
6.2-c : putting things away at the top modeler level

flomnes committed:
6.2c - simplify fixtures (#3188)

Use static methods instead of an interface class, remove need for
`initialize` and `virtual` functions.

Commit:
6.2-c : cleaning

Commit:
6.2-c : defining functions getDimensions(ctx)

Commit:
6.2-c : remove cumbersome namespace expansion

Commit:
6.2-c : simplify class VariablesBulkAddition

Commit:
6.2-c : oops, correction

Commit:
6.2c : oops, fix

Commit:
6.2-c : preparing to create and build master problem

Commit:
6.2-c : fix

Commit:
6.2-c : a bit of cleaning

Commit:
6.2-c : a bit of cleaning

Commit:
6.2-c : renaming

Commit:
6.2-c : creating a master and sub problem variables collector and passing it to component fillers

Commit:
[skip ci] 6.2-c : code formatting

Commit:
Merge remote-tracking branch 'remotes/origin/develop' into feature/6.2c--step-1

Commit:
6.2c - updates libs with config : variables

Commit:
6.2c - updates libs with config : remove things now useless

Commit:
6.2c - updates libs with config : remove things now useless

Commit:
6.2c - updates libs with config : remove things now useless

flomnes committed:
[6.2] Simplify location filtering for variable & objectives (#3200)

Commit:
6.2c - update libs with config : objectives

Commit:
6.2c - update libs with config : move optimConfig.h et consequences

Commit:
6.2c - update libs with config : fix

Commit:
6.2c - update libs with config : changes due to review

Commit:
6.2c - update libs with config : turn a member function into a free function

Commit:
6.2c - update libs with config : fixes due to review

guilpier-code committed:
6.2-c : fixes on optim-config.yml (#3202)

Co-authored-by: guilpier <guillaume.pierre_externe@rte-france.com>

Commit:
6.2c - update libs with config : cleaning

Commit:
Merge remote-tracking branch 'remotes/origin/feature/6.2c--step-2' into feature/6.2c--step-3

guilpier-code committed:
6.2-c : Fixes on objective contributions (#3203)

Asked by @tbittar

Co-authored-by: guilpier <guillaume.pierre_externe@rte-france.com>

Commit:
6.2-c : Fixes on objective contributions

Commit:
fixes on PR 3190 : some fix

Commit:
fixes on PR 3190 : avoid a unnecessary copy

flomnes commented on src/solver/optim-model-filler/ComponentFiller.cpp:
```diff
@@ -182,12 +226,19 @@ void VariablesBulkAddition::addVariable(const std::string& compoId,
     }
 }
 
+auto subproblemVariables = [](const Variable& v) { return v.isInSubProblem(); };
```
I see many conflicts with downstream branches

flomnes reviewed (commented):
The whole package seems to be getting ready, I think it's best to start merging

Commit:
fixes on PR 3190 : just cleaning a bit

Commit:
fixes on PR 3190 : changes due to review of #3190

guilpier-code committed:
6.2-c : update libraries with optim config (#3199)

Co-authored-by: guilpier <guillaume.pierre_externe@rte-france.com>
Co-authored-by: Florian Omnès <florian.omnes@rte-france.com>

flomnes committed:
Merge remote-tracking branch 'github/feature/6.2c--step-2' into fix/fixes-on-6.2-c--step-3--pr-3190

flomnes committed:
clang-format

flomnes committed:
Update src/solver/optim-model-filler/ComponentFiller.cpp

flomnes committed:
Use std::ranges::any_of

flomnes committed:
Add MasterAndSubPbVariables::write

flomnes committed:
Modeler logic

flomnes committed:
Skip constraints & objectives in master

Commit:
fixes on PR 3190 : fix a crash

Commit:
fixes on PR 3190 : fix another crash

flomnes committed:
Filter vars in OptimEntityContainer::addFromSystemComponents

flomnes committed:
Write master.mps, 1-1.mps in modeler mode

flomnes committed:
Fix additional variables

flomnes committed:
Fix logic

flomnes committed:
Reserve + comments

Commit:
Merge remote-tracking branch 'remotes/origin/fix/6.2c-modeler-final-steps' into fix/fixes-on-6.2-c--step-3--pr-3190

guilpier-code committed:
6.2-c : improve adding variables to problem (#3214)

Co-authored-by: guilpier <guillaume.pierre_externe@rte-france.com>

Commit:
6.2-c : small improvement

flomnes committed:
Remove some functions

flomnes committed:
Filter

flomnes committed:
Rename

flomnes committed:
if

flomnes committed:
Merge branch 'fix/6.2c-modeler-final-steps' into fix/fixes-on-6.2-c--step-3--pr-3190

Commit:
6.2-c : renaming due to review

guilpier-code committed:
6.2-c : printing info on subproblem into structure.txt  (#3190)

Co-authored-by: guilpier <guillaume.pierre_externe@rte-france.com>
Co-authored-by: Florian Omnès <florian.omnes@rte-france.com>

flomnes committed:
Merge remote-tracking branch 'github/develop' into feature/6.2c--step-1

flomnes committed:
Rename Connexion -> Connection

flomnes committed:
Partial revert of opt_appel_solveur_lineaire.cpp

Commit:
Merge branch 'fix/fixes-on-6.2-c--step-3--pr-3190' into feature/6.2c--step-1

Commit:
Merge branch 'feature/6.2c--step-1-loc' into feature/6.2c--step-1
